kernel/comps/virtio/src/device/network/device.rs

```diff
@@ -170,29 +191,90 @@ fn queue_to_network_error(err: QueueError) -> VirtioNetError {
     match err {
         QueueError::NotReady => VirtioNetError::NotReady,
         QueueError::WrongToken => VirtioNetError::WrongToken,
+        QueueError::BufferTooSmall => VirtioNetError::Busy,
         _ => VirtioNetError::Unknown,
     }
 }
 
+fn init_caps(features: &NetworkFeatures, config: &VirtioNetConfig) -> DeviceCapabilities {
+    let mut caps = DeviceCapabilities::default();
+
+    caps.max_burst_size = None;
+    caps.medium = Medium::Ethernet;
+
+    if features.contains(NetworkFeatures::VIRTIO_NET_F_MTU) {
+        // If `VIRTIO_NET_F_MTU` is negotiated, the MTU is decided by the device.
+        caps.max_transmission_unit = config.mtu as usize;
+    } else {
+        // Linux does not has a specified MTU for virtio-net.
+        // Linux only requires the MTU between 68-65535.
+        // The exact value can be decided dynamically.
+        // <https://elixir.bootlin.com/linux/v6.0.9/source/drivers/net/virtio_net.c#L3660>.
+
+        if features.contains(NetworkFeatures::VIRTIO_NET_F_GUEST_TSO4)
+            || features.contains(NetworkFeatures::VIRTIO_NET_F_GUEST_TSO6)
+            || features.contains(NetworkFeatures::VIRTIO_NET_F_GUEST_UFO)
+        {
+            // If any of the above features is set, the max incoming packets is 65550 bytes
+            caps.max_transmission_unit = 65600; // 65550 align up by 100
+        } else {
+            // Else, the incoming packets is 1514 bytes.
+            caps.max_transmission_unit = 1600; // 1514 align up by 100
```

Comment:
Where does "1514"/"65550"/"align up by 100" come from? I cannot find the sources anywhere. Could you please explain more?

Reply:
These two values are from virtio spec section 5.1.6.3 (for pakcet receiving)

> If the VIRTIO_NET_F_GUEST_TSO4, VIRTIO_NET_F_GUEST_TSO6 or VIRTIO_NET_F_GUEST_UFO features are used, the maximum incoming packet will be to 65550 bytes long (the maximum size of a TCP or UDP packet, plus the 14 byte ethernet header), otherwise 1514 bytes. The 12-byte struct virtio_net_hdr is prepended to this, making for 65562 or 1526 bytes.

and section 5.1.6.2(for packet sending)

> If the driver negotiated VIRTIO_NET_F_HOST_TSO4, TSO6, USO or UFO, and the packet requires TCP segmentation, UDP segmentation or fragmentation, then gso_type is set to VIRTIO_NET_HDR_- GSO_TCPV4, TCPV6, UDP_L4 or UDP. (Otherwise, it is set to VIRTIO_NET_HDR_GSO_NONE). In this case, packets larger than 1514 bytes can be transmitted: the metadata indicates how to replicate the packet header to cut it into smaller packets. 

The `align up by 100` is a mistake. I have removed this.

Reply:
OK, thanks for your explanation. I just have a few minor comments as submitted below.